streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
@mjsax if you are talking about a real hot code path, then even assertions can have an impact because they increase bytecode size, which can prevent inlining by HotSpot. Anyway, you are advocating a new style that is not used anywhere else in Kafka. I don't think this PR is the right venue for that discussion.